### ISSUE TITLE
Setting struct members of a C++ class via Lua

### ISSUE THREAD

**Issue by javawag** [vinniefalco/LuaBridge#128]
Hi, 

I think I've hit a limitation of LuaBridge but I just wanted to check I'm not doing something silly.

Given this class mapping in C++:

``` cpp
luabridge::getGlobalNamespace()
    .beginNamespace("engine")
        .beginClass<vec3>("vec3")
            .addConstructor<void(*)(float, float, float)>()
            .addData("x", &vec3::X)
            .addData("y", &vec3::Y)
            .addData("z", &vec3::Z) 
        .endClass()
        .beginClass<GameObject>("GameObject")
            .addData("position", &GameObject::Position)
        .endClass()
    .endNamespace();
```

Now from Lua I want to set a GameObject's x position, i.e. go.Position.X in C++, or go.position.x in Lua.

This works: `go.position = engine.vec3(go.position.x, 2000, go.position.z)` but is clearly very verbose. However, `go.position.y = 2000` does not work - it simply does nothing. Calling `print(go.position.y)` before and after setting the value in Lua shows the old value on both calls, too.

Any pointers? It appears that go.position is returning a copy or something, and so the changes don't affect the original variable.

**Comment by sillykelvin** [comment]
I met the same problem, Is there anyway or workaround to pass the struct member by pointer or by reference but not by value? I checked the manual, it does not mention this.

**Comment by javawag** [comment]
It's been a *really* long time since I did anything with this, but I think I semi-answered my own question. Note that I stated in the question that it "seemed" like it was returning a copy (it was!).

I believe my solution was (e.g. for `position` above), to add a property using two functions as below:

```
void Lua_setPosition(vec3* newPosition) {
   Position = *newPosition;
}
vec3* Lua_getPosition() {
   return &Position;
} 
```

See the docs for `addProperty` to see how you hook these up. Note that these take a pointer to a `vec3`, which avoids taking a copy. Inside the function I then dereference or take the address accordingly. I then made a macro so that I could do the above more easily.

Hope that helps!

**Comment by sillykelvin** [comment]
Wow, thanks for your quick response, I dived into the source code, and seems it DOES return a copy: https://github.com/vinniefalco/LuaBridge/blob/master/Source/LuaBridge/detail/Userdata.h#L400

And thanks for your workaround, I solved the problem like that. However, IMO, both `go.position = Vector3(1, 1, 1)` and the `addProperty` approach do not solve the root problem, due to:

1. `go.position.x = 1` is so straightforward and we easily write code like that
2. there is neither compiling error nor runtime error
3. it's hard to debug, I never realized the bug is because `go.position` returns a copy
4. even if I know the behavior now, I have to remember it all the time when I export C++ classes to Lua, and have to write more glue code to avoid it
5. newbies might make the same mistake

So, I don't know if there is any better solution, IMO, a better solution is: `go.position.x = 1` can work as expected; or, at least, give a compiling error or runtime error/warning.

@dmitry-t @vinniefalco Any comments on this issue?

**Comment by javawag** [comment]
No worries at all - glad I could help at least somewhat!

I'm not sure, but I think I solved the `go.position.x" problem by using a similar hack in my `vec3` class - i.e. adding `getX/Y/Z()` and `setX/Y/Z()` functions taking pointers. I think.

Maybe vinnie has a better suggestion though, as it was just a hack!

**Comment by dmitry-t** [comment]
@sillykelvin Indeed, `position` returns a copy. So far you can use the workaround proposed by @javawag - explicitly return a pointer.

**Comment by dmitry-t** [comment]
This should have been fixed by #189.
